# Introduction to Programming

Purpose of the chapter: introduce using python as a computing tool.
- The command line: computing by hand: walk through computing
$F_{10}$ the stupidest way possible.
- Loops: for loops, while loops: using these to compute $F_{10}$ in
a slightly less stupid way.
- Lists: saving computations, plotting on regular scale, plotting on
log scale: conjectures and
questions.
- Functions: if something is done over and over again, make it a
function: inputs, outputs
- Recursion: the danger of simplistic recusion: why computing
Fibonacci numbers recursively can
lead to exponentially many sub-calls: do we talk about decorations and
how to avoid this?
- Complexity of computation: time complexity, space complexity: both
impact computation of
and display of big Fibonacci numbers
- The idea that just because an algorithm exists doesn't make it the
{\em best} algorithm.
- Better algorithms for Fibonacci numbers?
- Better algorithms for just the last few digits: iterated squaring,
intro to $2\times 2$ matrix
computations.
- Experiments to perform: Pisano periods



## The Fibonacci Numbers
The Fibonacci numbers, $F_n$, named after the nickname given to Leonardo Pisano, or 
Leonardo of Pisa,  were actually studied as early as 200 BCE by Pingala, who used them to analyse 
the number of patterns in Sanskrit poetry.  They are define by a two-term linear recurrence relation,

$$F_{n}= F_{n-1} + F_{n-2}$$

and two initial conditions.  There are several different conventions about what the initial 
conditions should be: we will choose $F_0=0$ and $F_1=1$.  In this section, we'll explore how to 
compute the values taken by $F_n$ for various different $n$: we'll start by doing some really 
stupid things, and gradually develop better methods.  If you have lots of coding experience, you
will recognize the way we are doing things as stupid: but if you are new to coding, we hope that this
will gently get you started to thinking a computational way.

We'll start with variables: in Python, variables are objects that can store a value: in our cases, the
values will be integers.  In the Jupyter notebook, input the following code.

In [19]:
F0=0
F1=1
F2=F1+F0
print(F2)

1


Execute the code by clicking the run arrow (or typing shift-enter)
and you should see the output 1, indicating that you have successfully printed $F_1$.

In the next code block, type 

In [4]:
F3=F2+F1
print(F3)
F4=F3+F2
print(F4)
F5=F4+F3
print(F5)

2
3
5


By now this should be feeling pretty repetitive, and we should be longing for a way to not have to keep typing
two more rows of code for each new Fibonacci number.  Let's try to be more systematic about it.  Let's think about what we are doing: we have a current Fibonacci number, a previous Fibonacci number, and we use them to compute the 
next one in the sequence.

$$next = current + previous.$$

Then we replace the previous one by the one which was current, and the current by the next, and repeat.

In [5]:
previous =0
current = 1
next = current + previous
previous = current
current = next
next = current + previous
previous = current 
current = next
next = current + previous
previous = current 
current = next
next = current + previous
previous = current 
current = next
print(next)


5


### Loops


This may seem worse than the previous code: three lines of code for each step, but since we are no longer using a silly naming scheme for our variables, we can use loops to automate things.  To do this we can use for or while loops.

In [8]:
previous = 0
current = 1
for i in range(2,6):
    next = current + previous
    previous = current
    current = next
print(i,next)    

5 5


Now we see that we can modify our loop to make it compute much further out:

In [10]:
previous = 0
current = 1
for i in range(2,31):
    next = current + previous
    previous = current
    current = next
print(i,next)   

30 832040


This tells us that $F_{30}= 832040$.

We can also use while loops to do something similar:


In [11]:
previous = 0
current = 1 
i=2 
while i< 31:
    next = current + previous
    previous = current
    current = next
    i+=1
print(i,next)   

31 832040


In this, we need to set the value of the variable: as long as the value of `i` is less than 31, the loop will be executed: in this example we have to remember to increment the value of `i` by 1 each time we execute the loop.

### Lists

Suppose that we now want to keep our computed values around once we have computed them.  One way of achieving this is to put the values in a list.

In [13]:
fib_list = [0 , 1]  # We populate the list with our initial values F_0=0, F_1=1
for i in range(2,11):
    previous = fib_list[i-2]
    current= fib_list[i-1]
    next = current+previous
    fib_list.append(next)
print(fib_list)  
print(fib_list[5])

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55]
5


Perhaps we wish instead to create a list of pairs, $[i,F_i]$: 

In [20]:
fib_list = [[0,0] , [1, 1]]  # We populate the list with our initial values F_0=0, F_1=1
for i in range(2,11):
    previous = fib_list[i-2][1]
    current= fib_list[i-1][1]
    next = current+previous
    fib_list.append([i,next])
print(fib_list)  
print(fib_list[5])

[[0, 0], [1, 1], [2, 1], [3, 2], [4, 3], [5, 5], [6, 8], [7, 13], [8, 21], [9, 34], [10, 55]]
[5, 5]


### Functions


If we wish to compute lots of values of Fibonacci numbers there are various things we could do: precompute all the values we need and store them in a list, for example.  But often the best thing to do is to write a function which we can call when we need a particular value.


In [14]:
def fibo(n):
    previous = 0
    current = 1 
    i=2 
    while i< n+1:
        next = current + previous
        previous = current
        current = next
        i+=1
    return(next)    
print(fibo(5)) 
print(fibo(30))
    

5
832040


We can now test how fast our code is: for how large an input $n$ can we compute `fibo(n)` in realistic time?
As an aside, in our context we use "realistic time" to mean a few seconds or a minute.  Depending on the importance of a computation, realistic time could mean minutes, months, or even years.

In [15]:
import time
t=time.time()
fibo(1000)
time.time()-t

0.0009942054748535156

In [16]:
t=time.time()
fibo(10000)
time.time()-t

0.023461580276489258

In [17]:
t=time.time()
fibo(100000)
time.time()-t

0.3279304504394531

In [18]:
t=time.time()
fibo(1000000)
time.time()-t

10.424722671508789

We see that by the time we get to $n=10^6$, the code has started to slow down considerably.

### Complexity in time and space


We need to think about how many steps a program will take, and how long each step will take.  For the function `fibo()` it is easy to see that computing `fibo(n)` will take $n-1$ additions: however, this is not the end of the 
story, since the summands are increasing in size, and adding large numbers takes longer than adding small numbers.
How do we measure the size of an input $n$?  The best way to think of this is how many bits are needed to specify  $n$: that is, about $\log_2(n)$.  

Now the number of bitwise operations needed to add two $k$-bit integers grows linearly in $k$.  Let's suppose that it takes $c_1 k$ bitwise operations.

###  Exercise 

Compute the $n$ Fibonacci number for $n=100$, $1000$, and $10000$.  How many digits does each have?
Can you predict the number of digits in $F_n$ as a function of $n$?  How many bits in the binary expansions?

Explore Binet's formula for $F_n$: this should allow you to obtain a formula for the number of bits in $F_n$.
Let's suppose that $F_n$ has $c_2 n$ bits.


Back to complexity: adding $F_{n-1}$ and $F_{n-2}$ involves summing two numbers with $c(n-1)$ and $c (n-2)$ bits: this takes $c_1 c_2 (n-1)$ bitwise operations.  If we have to do this for all the smaller Fibonacci numbers too,
we would end up with 

$$\sum_{k<n}c_1 c_2 k=c_1 c_2\binom{n}{2}$$

This suggests that the time complexity of the function `fibo(n)` is quadratic in $n$ (_not the size of $n$_): that is if we double $n$, we roughly quadruple the time to compute `fibo(n)`.

The space complexity of this algorithm is easier to compute: we basically need a constant multiple of the number 
of bits of output, which we saw earlier was $c_1 n$.  We should note that this, while linear in $n$, is exponential in $\log_2 n$, the _size_ of $n$.

### Recursion

We can sometimes write simpler code if we use the idea of recursion.  Let's define a function `fibor(n)` which will recursively call `fibor(n-1)` and `fibor(n-2)`.

In [23]:
def fibor(n):
    if n== 0:
        return(0)
    elif n==1:
        return(1)
    else:
        return(fibor(n-1)+fibor(n-2))
print(fibor(5))
print(fibor(30))

5
832040


We should point out, however, that as intuitive and nice as recusive code can be, issues can arise.  Without care, and especially in a case like this, recursive function calls can grow exponentially.  

In [39]:
counter=0  # This will count the number of times we call fibor
def fibor(n):
    global counter # This will allow fibor() to reference the variable counter inside the function
    counter+=1
    if n== 0:
        return(0)
    elif n==1:
        return(1)
    else:
        return(fibor(n-1)+fibor(n-2))
print(fibor(5))
print(counter)
counter=0
print(fibor(30))
print(counter)

5
15
832040
2692537


We see that in computing $F_5$ we already made 15 calls to `fibor()`: in computing $F_{30}$ we made over 2 million recursive calls.  This means that the recursive function `fibor()` is unusable for even moderate values of $n$.


In [40]:
counter=0
t=time.time()
print(fibor(30))
print(counter)
print(time.time()-t)
counter=0
t=time.time()
print(fibor(40))
print(counter)
print(time.time()-t)

832040
2692537
0.3839454650878906
102334155
331160281
51.42229461669922


The problem here is this: considering $F_5$ for concreteness: $F_5$ calls $F_4$ and $F_3$.  $F_4$ calls $F_3$ and $F_2$.  Each of the two $F_3$ calls calls $F_2$ and $F_1$.  Each of the three $F_2$ calls calls $F_1$ and $F_0$: 
thus we get 
$$1 (F_5) + 1(F_4) + 2(F_3) + 3(F_2) + 5(F_1)+3(F_0) = 15 $$
total calls to `fibor()`.  

There are ways to avoid this, using "memoization" and "decorators" to ensure that python realizes that we have already called $F_k$ and doesn't spawn multiple subroutines calling the same function, but that is a more 
advanced topic than we are covering here.

### Is there a better way?


One of the themes that we will stress along the way is that, just because we have an algorithm or an implentation 
of an algorithm which works, doesn't mean that we should stop there: there may be a better algorithm.  Perhaps one that runs faster, or uses less memory.  Sometimes we'll be able to find algorthms that will run _much_ faster, or use much less space!  Sometimes we can trade off using more space to use much less time.

### Matrices and linear recurrences

Let's observe that we can rewrite the Fibonacci recurrence as a linear matrix-vector recurrence in a very simple way.  
If we let $\underline{v}_n$ denote the vector

$$
\underline{v}_n = 
\begin{bmatrix}  F_n\\ F_{n-1}
\end{bmatrix}
$$

and let $A$ denote the matrix 

$$
A=\begin{bmatrix}
1 & 1 \\
1 & 0
\end{bmatrix}
$$

then we can rewrite the Fibonacci recurrence as 

$$\underline{v}_n = A \underline{v}_{n-1}$$

which can be followed back to deduce that 

$$
\underline{v}_n = A^n \underline{v}_0.
$$

This means that if we can compute powers of $A$ efficiently we can also compute $F_n$ efficiently.
This brings us to the following important idea.

### Iterated squaring to compute powers 

Consider the expression $A^{25}$.  Naively, multiplying 25 copies of $A$ together takes 24 multiply operations: this can be significantly improved as follows: compute $A$, $A^2$, $A^4=(A^2)^2$, $A^8=A^4)^2$ and $A^{16}=(A^8)^2$.  This takes 4 multiplies.  Now compute $25=16+8+1$, so $A^{25}=A^{16}A^8A$ requiring two more multiplies, for a total of 6.

In general, to compute $A^n$, we compute $A^{2^k}$ for $k$ up to $\log_2 n$ using iterated squaring.  
Then we multiply the appropriate subset of these iterated squares together by computing the binary expansion of $n$.  This takes a total of fewer than $2 \log_2 n$ matrix multiplications.

Now, we have replaced the complexity of adding enormous numbers with the complexity of multiplying enormous numbers, but far fewer times.  If you choose to study the complexity of multiplying two large integers together, there is a wonderful rabbit hole to go down, starting from naive multiplication of two $n$ bit integers, which 
takes $O(n^2)$ operations, to the beautiful Karatsuba' algorithm, with complexity about $O(n^{1.585})$, to the 
Schönhage-Strassen algorithm which has asymptotic complexity $O(n\log n \log\log n)$, and the recent 
$O(n \log n)$ algorithm of Harvey and van der Hoeven.

If we use these more sophisiticated methods, we can obtain an algorithm for computing $F_n$ of order about 
$O(n \log n)$,  We can't expect much better than that, since even outputting the digits of $F_n$ is of order 
$O(n)$.

### Exercise: 
Figure out how to implement iterated squaring to compute $A^51$ by a sequence of python statements.  Then code a function which will take $n$ as an input, and return $A^n$.  You should only need to keep one expression of the 
form $A^{2^k}$ around at a time if you are careful.  Hint: find the low order bits of $n$ first.

it is clear at this point that what limits our computation of big Fibonacci numbers is the size of the results. What if we only want to compute the final few digits of $F_n$?  Now the size of the output (and of the intermediate computations) is kept small, and the limiting factor is the number of squarings of matrices we are going to perform.  Consequently it is now feasible to take something like $n=7^{12345678}$ and compute the 
final 100 digits of $F_n$.



### Again, is there a better way?
We can revisit the question: is there a better way to compute this?